File: 2024-12-December-GFG-POTD/December 27 - Count pairs with given sum/Solution.java
import java.io.*;
import java.lang.*;
import java.util.*;

class Solution {

    int countPairs(int arr[], int target) {
        // Create a hash map to store frequencies of elements
        Map<Integer, Integer> freqMap = new HashMap<>();
        int count = 0;

        // Traverse the array
        for (int num : arr) {
            int complement = target - num;

            // If the complement is already in the map, increment the count
            if (freqMap.containsKey(complement)) {
                count += freqMap.get(complement);
            }

            // Add the current number to the map
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        return count;
    }
}

// Driver code
class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());

        while (t-- > 0) {
            String line = read.readLine().trim();
            String[] numsStr = line.split(" ");
            int[] nums = new int[numsStr.length];
            for (int i = 0; i < numsStr.length; i++) {
                nums[i] = Integer.parseInt(numsStr[i]);
            }

            int target = Integer.parseInt(read.readLine());

            Solution obj = new Solution();
            System.out.println(obj.countPairs(nums, target));
            System.out.println("~");
        }
    }
}

File: 2024-12-December-GFG-POTD/December 27 - Count pairs with given sum/Solution.py
class Solution:
    def countPairs(self, arr, target):
        freq_map = {}  # Dictionary to store frequencies of elements
        count = 0

        # Traverse the array
        for num in arr:
            complement = target - num

            # If the complement is already in the map, increment the count
            if complement in freq_map:
                count += freq_map[complement]

            # Add the current number to the map
            freq_map[num] = freq_map.get(num, 0) + 1

        return count

# Driver code
if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        nums = list(map(int, input().split()))
        target = int(input())

        obj = Solution()
        print(obj.countPairs(nums, target))
        print("~")
